File: typescript/packages/core/tools/matimo_search_tools/matimo_search_tools.ts
import { getGlobalMatimoInstance, getGlobalMatimoLogger, validateToolDefinition, classifyRisk } from '@matimo/core';
import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';

interface SearchParams {
  query: string;
  limit?: number;
}

interface ToolSummary {
  name: string;
  description: string;
  version: string;
  tags: string[];
  riskLevel: string;
}

interface SearchResult {
  results: ToolSummary[];
  total: number;
  query: string;
}

export default async function matimoSearchTools(params: SearchParams): Promise<SearchResult> {
  const logger = getGlobalMatimoLogger();
  const query = params.query ?? '';
  const limit = params.limit ?? 20;

  const toSummary = (tool: { name: string; description: string; version: string; tags?: string[] }): ToolSummary => ({
    name: tool.name,
    description: tool.description,
    version: tool.version,
    tags: tool.tags ?? [],
    riskLevel: classifyRisk(tool as Parameters<typeof classifyRisk>[0]),
  });

  // Prefer registry search via the global instance (has all loaded tools)
  try {
    const instance = getGlobalMatimoInstance();
    const found = instance.searchTools(query).slice(0, limit);
    logger.debug('matimo_search_tools: registry search', { query, count: found.length });
    return {
      results: found.map(toSummary),
      total: found.length,
      query,
    };
  } catch {
    // Global instance not set — fall through to disk-based scan below
    logger.debug('matimo_search_tools: no global instance, falling back to disk scan');
  }

  // Fallback: scan the default tools directory
  const toolDir = './matimo-tools';
  if (!fs.existsSync(toolDir)) {
    return { results: [], total: 0, query };
  }

  const lowerQuery = query.toLowerCase();
  const results: ToolSummary[] = [];

  for (const entry of fs.readdirSync(toolDir, { withFileTypes: true })) {
    if (!entry.isDirectory()) continue;
    const defPath = path.join(toolDir, entry.name, 'definition.yaml');
    if (!fs.existsSync(defPath)) continue;

    try {
      const tool = validateToolDefinition(yaml.load(fs.readFileSync(defPath, 'utf-8')));
      const matches =
        tool.name.toLowerCase().includes(lowerQuery) ||
        tool.description.toLowerCase().includes(lowerQuery) ||
        (tool.tags ?? []).some((t) => t.toLowerCase().includes(lowerQuery));

      if (matches) results.push(toSummary(tool));
    } catch {
      // Skip invalid definitions silently
    }
    if (results.length >= limit) break;
  }

  return { results, total: results.length, query };
}

File: python/packages/core/src/matimo/tools/matimo_search_tools/matimo_search_tools.py
"""matimo_search_tools — search the loaded tool registry by keyword."""
from __future__ import annotations

import logging
from pathlib import Path

import yaml

logger = logging.getLogger("matimo")


async def run(params: dict) -> dict:  # type: ignore[type-arg]
    from matimo.core.models import ToolDefinition
    from matimo.policy.risk_classifier import classify_risk

    query: str = params.get("query", "")
    limit: int = int(params.get("limit", 20))

    def to_summary(tool: ToolDefinition) -> dict:  # type: ignore[type-arg]
        try:
            risk = classify_risk(tool)
        except Exception:
            risk = "medium"
        return {
            "name": tool.name,
            "description": tool.description,
            "version": tool.version,
            "tags": list(tool.tags or []),
            "riskLevel": risk,
        }

    # Prefer global instance (has all loaded tools)
    try:
        from matimo.decorators import get_global_matimo_instance

        instance = get_global_matimo_instance()
        if instance is not None and hasattr(instance, "search_tools"):
            found = instance.search_tools(query)[:limit]
            logger.debug("matimo_search_tools: registry search query=%s count=%d", query, len(found))
            return {"results": [to_summary(t) for t in found], "total": len(found), "query": query}
    except Exception as exc:
        logger.debug("matimo_search_tools: no global instance, falling back to disk: %s", exc)

    # Fallback: scan default tools directory
    tool_dir = Path("./matimo-tools")
    if not tool_dir.exists():
        return {"results": [], "total": 0, "query": query}

    lower_query = query.lower()
    results: list[dict] = []  # type: ignore[type-arg]

    for entry in tool_dir.iterdir():
        if not entry.is_dir():
            continue
        def_path = entry / "definition.yaml"
        if not def_path.exists():
            continue
        try:
            parsed = yaml.safe_load(def_path.read_text(encoding="utf-8"))
            tool = ToolDefinition.model_validate(parsed)
            tags_lower = [t.lower() for t in (tool.tags or [])]
            if (
                lower_query in tool.name.lower()
                or lower_query in tool.description.lower()
                or any(lower_query in t for t in tags_lower)
            ):
                results.append(to_summary(tool))
        except Exception as exc:
            logger.debug("matimo_search_tools: skipping invalid tool %s: %s", entry.name, exc)
        if len(results) >= limit:
            break

    return {"results": results, "total": len(results), "query": query}
